en/_posts/2024-05-23-understanding-github-artifact-attestations.md

```diff
@@ -0,0 +1,224 @@
+understand a bit about how GitHub Artifact Attestations work and their relation
+to SLSA levels.
+
+## Architecture
+
+Generating attestations is done using the
+[`attest-build-provenance`](https://github.com/actions/attest-build-provenance)
+GitHub action. Github’s blog post does a good job of explaining how it works so
+I won’t rehash it fully here. I’ll just summarize the flow and highlight some
+additional information that will be important later.
+
+1. `attest-build-provenance` requests an OIDC token from the GitHub OIDC
+   provider. This OIDC token contains [information about the
+   build](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)
+   in the token claims.
+
+2. The OIDC token is sent to a [Sigstore
+   Fulcio](https://github.com/sigstore/fulcio) server (either the public instance
+   or GitHub’s private one). [Fulcio can recognize and validate GitHub’s OIDC
+   tokens](https://docs.sigstore.dev/certificate_authority/oidc-in-fulcio/#github)
+   and, after verifying its signature, issues a certificate in exchange for the
+   OIDC token. This certificate includes much of the information from the OIDC
+   token [mapped into its OID extension
+   fields](https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#mapping-oidc-token-claims-to-fulcio-oids)
+   as claims.
+
+3. SLSA Provenance is generated and the resulting JSON is signed with the
+   returned certificate’s private key. This private key is then discarded.
+   The provenance is recorded in GitHub’s attestation store.
+
+I'm leaving out some details but this is the general flow for attestation. So
+as a result we have some provenance in JSON format and a Sigstore certificate
+with some OID claims set.
+
+After we have an artifact and attestation, as a user, we need to be able to
+verify it before we use it. Verification works something like the following.
+The code for this is found in the [`gh attestations verify
+command`](https://github.com/cli/cli/tree/trunk/pkg/cmd/attestation/verify)
+for GitHub’s official CLI tool.
+
+1. The attestation is downloaded from the GitHub Attestations API using the
+   artifact’s digest.
+2. The attestation’s signature is verified against the public key used to sign
+   it, and to the certificate chain for either the GitHub Fulcio instance, or
+   the public Sigstore Fulcio instance (the root certificates are managed using
+   [TUF](https://theupdateframework.io/)).
+3. The expected values for the owner or repo given by the user are matched
+   against the signing certificate’s OID claims.
+
+Notice that nowhere here did we actually use the contents of the SLSA
+provenance for verification. We’ll discuss why this is below.
+
+## A Good User Experience
+
+By providing a GitHub Action, GitHub gives users the maximum amount of
+flexibility when integrating this into their GitHub Actions workflows. It’s
```

Comment:
something missing today is how to call this programatically from an API, eg from a package manager CLI. They should turn their Action into a sidecar that runs locally on the runner, so that package managers can simplify fo a REST call to get an attestation.

Reply:
Right. For example, they built the logic into npm when it probably could have had it call this API to get an attestation before publishing to the npm registry.

Reply:
They do have an [`attest` action](https://github.com/actions/attest) which can be used to get an attestation. It generates and returns a path to a bundle file.

Reply:
This is something to note but I didn't add anything to the post since maybe too far out of scope.